### PR TITLE
UDF for Variance

### PR DESCRIPTION
**Pull request by evionkim**
Please review these. Thanks a lot!

Please ignore comments below. Works just fine after ant clean & rebuild.
## 

Found out that tests are failing from error: " java.lang.RuntimeException: Found more JAR files than expected" , after merging with master. 

It seems like other tests are failing from same error, too.

### CODE REVIEW

File: src/java/datafu/pig/stats/VAR.java [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+import org.apache.pig.data.DataType;
+import org.apache.pig.data.Tuple;
+import org.apache.pig.data.TupleFactory;
+import org.apache.pig.impl.logicalLayer.FrontendException;
+import org.apache.pig.impl.logicalLayer.schema.Schema;
+import org.apache.pig.backend.executionengine.ExecException;
+
+
+/**
+* Generates the {@link <a href="http://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval#Wilson_score_interval" target="_blank">Variance/a>}
```
Comment (samshah) [inline src/java/datafu/pig/stats/VAR.java]: Incorrect wikipedia link

File: src/java/datafu/pig/stats/VAR.java [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+import org.apache.pig.impl.logicalLayer.schema.Schema;
+import org.apache.pig.backend.executionengine.ExecException;
+
+
+/**
+* Generates the {@link <a href="http://en.wikipedia.org/wiki/Variance" target="_blank">Variance/a>} 
+* of a set of Values. This UDF uses the fact that variance(x) = average(x^2) - average(x)^2
+* This class implements * {@link org.apache.pig.Algebraic}, so if possible the execution will performed in a distributed fashion.
+* VAR implements the {@link org.apache.pig.Accumulator} interface as well.
+*/
```
Comment (samshah) [inline src/java/datafu/pig/stats/VAR.java:45]: Add example of use too please

File: src/java/datafu/pig/stats/VAR.java [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+
+    @Override
+    public Schema outputSchema(Schema input) {
+        return new Schema(new Schema.FieldSchema(null, DataType.DOUBLE));
+    }
+
+    @Override
+    public List<FuncSpec> getArgToFuncMapping() throws FrontendException {
+        List<FuncSpec> funcList = new ArrayList<FuncSpec>();
+        funcList.add(new FuncSpec(this.getClass().getName(), Schema.generateNestedSchema(DataType.BAG, DataType.BYTEARRAY)));
```
Comment (samshah) [inline src/java/datafu/pig/stats/VAR.java:311]: Can we perhaps use generics instead of so much code duplication?

File: test/pig/datafu/test/pig/stats/VARTests.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+import static org.testng.Assert.*;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.pig.data.Tuple;
+import org.apache.pig.pigunit.PigTest;
+import org.testng.annotations.Test;
+
+import datafu.test.pig.PigTests;
```
Comment (samshah) [inline test/pig/datafu/test/pig/stats/VARTests.java:12]: Tests for corner cases (e.g., 0 elements, 1 element)

File: src/java/datafu/pig/stats/VAR.java (hunk 3 above)
Comment (evionkim) [inline src/java/datafu/pig/stats/VAR.java:311]: Can't because inner functions/inner classes are static.

File: test/pig/datafu/test/pig/stats/VARTests.java (hunk 4 above)
Comment (evionkim) [inline test/pig/datafu/test/pig/stats/VARTests.java:12]: Test added.

File: src/java/datafu/pig/stats/DoubleVAR.java [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+                // input is a bag with one tuple containing
+                // the column we are trying to get variance 
+                DataBag bg = (DataBag) input.get(0);
+                Double d = null;
+                if(bg.iterator().hasNext()) {
+                    Tuple tp = bg.iterator().next();
+                    d = (Double)tp.get(0);
+                }
+                t.set(0, d);
+                t.set(1, d*d);
```
Comment (matthayes) [inline src/java/datafu/pig/stats/DoubleVAR.java]: what if d is null?

File: src/java/datafu/pig/stats/DoubleVAR.java [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+
+        Tuple output = mTupleFactory.newTuple(3);
+        boolean sawNonNull = false;
+        for (Iterator<Tuple> it = values.iterator(); it.hasNext();) {
+            Tuple t = it.next();
+            Double d = (Double)t.get(0);
+            Double dSquare = (Double)t.get(1);
+            Long count = (Long)t.get(2);
+            
+            // we count nulls in avg as contributing 0
```
Comment (matthayes) [inline src/java/datafu/pig/stats/DoubleVAR.java]: update comment, not avg

File: test/pig/datafu/test/pig/stats/VARTests.java (hunk 4 above)
Comment (matthayes) [inline test/pig/datafu/test/pig/stats/VARTests.java:12]: Also add tests for null values in the data

File: test/pig/datafu/test/pig/stats/VARTests.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    List<Tuple> output = getLinesForAlias(test, "data_out", true);
+    
+    assertEquals(output.size(),1);
+    assertEquals(output.get(0).toString(), "(8.25)");
+  }
+  
+  @Test
+  public void varTestFloat() throws Exception
+  {
+    PigTest test = createPigTest("test/pig/datafu/test/pig/stats/varTestDouble.pig");
```
Comment (matthayes) [inline test/pig/datafu/test/pig/stats/VARTests.java]: varTestFloat

File: src/java/datafu/pig/stats/DoubleVAR.java (hunk 6 above)
Comment (samshah) [inline src/java/datafu/pig/stats/DoubleVAR.java]: I think nulls should be dropped (i.e., they are NAs), not counted as 0. Thoughts?

**Comment by samshah** [comment]
Additionally, add unit test for null entries.

File: src/java/datafu/pig/stats/DoubleVAR.java (hunk 6 above)
Comment (evionkim) [inline src/java/datafu/pig/stats/DoubleVAR.java]: From the class Initial, third entry of tuple(count) is set to zero if input is null, which in resul t is 'dropping' the null entry.
I'll commit my new unit test soon, however from the test,
Variance of (1,2,3,4,5,6,7,8,9,10) = Variance of(1,2,3,4,5,6,7,8,9,10,null).
